configfile: "config/config.yaml"

include: "rules/1_assembly.smk"
include: "rules/2_annotation.smk"
include: "rules/3_genome_level.smk"
include: "rules/4_homology_level.smk"
include: "rules/5_domain_level.smk"

rule all:
    input:
         #fastqc_before_trimming,
         "results/Genomics/1_Assembly/1_Preprocessing/fastqc_before_trimming",
         ##flye
         expand("results/Genomics/1_Assembly/2_Assemblers/flye/{process}",
               process=["raw"]),
         #blastn contamination detection
         #expand("results/Genomics/1_Assembly/3_Evaluation/blastn/{assembler}/assembly.blastn",
          #      assembler=["flye/raw"]),
          #winnowmap evaluation
         expand("results/Genomics/1_Assembly/3_Evaluation/winnowmap/{assembler}/merlyDB",
                assembler=[ "flye/raw"],
                genome=["barkhanus"]),
         expand("results/Genomics/1_Assembly/3_Evaluation/winnowmap/{assembler}/repetitive_k15.txt",
                assembler=[ "flye/raw"],
                genome=["barkhanus"]),
         expand("results/Genomics/1_Assembly/3_Evaluation/winnowmap/{assembler}/{long_reads}.bam",
                assembler=[ "flye/raw"],
                long_reads=["nanopore"]),
         #quast
         expand("results/Genomics/1_Assembly/3_Evaluation/quast/{assembler}/",
                assembler=[ "flye/raw"]),
        #deeptools
         expand("results/Genomics/1_Assembly/3_Evaluation/deeptools/{assembler}.png",
                assembler=[ "flye/raw"]),
        #multiqc
         expand("results/Genomics/1_Assembly/3_Evaluation/multiqc/{assembler}",
          assembler=[ "flye/raw"]),
         #prodigal
         expand("results/Genomics/2_Annotation/1_Structural/prodigal/{assembler}/genome.gff",
                assembler=["flye/raw"]),
         #glimmerhmm
        # expand("results/Genomics/2_Annotation/1_Structural/glimmerhmm/{assembler}/genome.gff",
         #  assembler=["flye/raw"])



